File: tools/extract_defs.c
/*
 * extract_defs.c — Extract math variable definitions from parse trees.
 *
 * Usage: extract_defs file.parsed.txt [sidecar.tok]
 *
 * Reads constituency parse trees (one ROOT per sentence), finds CD
 * nodes containing numeric token IDs, matches syntactic patterns to
 * extract the defining noun phrase, and outputs TSV:
 *   token_id \t math_text \t definition
 *
 * If sidecar.tok is provided, token IDs are mapped to their original
 * math text (e.g. "$\alpha$") for the second column.
 */
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>
#include "parse_tree.h"

/* from match_patterns.c */
int  match_definition(tree_t *t, int cd_idx, char *buf, int bufsz);
bool n_is_cd_id(tree_t *t, int idx);

/* ---- sidecar.tok parser ---- */
typedef struct {
    char *token_id;
    char *token_text;
} idmap_entry_t;

typedef struct {
    idmap_entry_t *entries;
    int n;
} idmap_t;

static idmap_t *load_sidecar(const char *path) {
    FILE *fp = fopen(path, "r");
    if (!fp) return NULL;
    idmap_t *m = calloc(1, sizeof(*m));
    m->entries = malloc(4096 * sizeof(idmap_entry_t));
    char line[4096];
    while (fgets(line, sizeof(line), fp) && m->n < 4096) {
        char *tid = strstr(line, "token_id:");
        char *tok = strstr(line, "<tok:");
        if (!tid || !tok) continue;
        tid += 9;  /* skip "token_id:" */
        char *tc = strchr(tid, ',');
        if (!tc) continue;
        *tc = '\0';
        tok += 5;  /* skip "<tok:" */
        char *te = strstr(tok, ">}");
        if (!te) { te = strchr(tok, '>'); }
        if (!te) continue;
        *te = '\0';
        m->entries[m->n].token_id = strdup(tid);
        m->entries[m->n].token_text = strdup(tok);
        m->n++;
    }
    fclose(fp);
    return m;
}

static char *idmap_lookup(idmap_t *m, const char *tid) {
    if (!m) return NULL;
    for (int i = 0; i < m->n; i++)
        if (!strcmp(m->entries[i].token_id, tid))
            return m->entries[i].token_text;
    return NULL;
}

static void free_idmap(idmap_t *m) {
    if (!m) return;
    for (int i = 0; i < m->n; i++) {
        free(m->entries[i].token_id);
        free(m->entries[i].token_text);
    }
    free(m->entries);
    free(m);
}

/* ---- read file ---- */
static char *read_file(const char *path, size_t *out_len) {
    FILE *fp = fopen(path, "r");
    if (!fp) return NULL;
    fseek(fp, 0, SEEK_END);
    long sz = ftell(fp);
    fseek(fp, 0, SEEK_SET);
    char *buf = malloc((size_t)sz + 1);
    size_t n = fread(buf, 1, (size_t)sz, fp);
    buf[n] = '\0';
    fclose(fp);
    if (out_len) *out_len = n;
    return buf;
}

/* ---- main ---- */
int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: extract_defs file.parsed.txt [sidecar.tok]\n");
        return 1;
    }

    idmap_t *idmap = NULL;
    if (argc >= 3)
        idmap = load_sidecar(argv[2]);

    size_t flen;
    char *text = read_file(argv[1], &flen);
    if (!text) { fprintf(stderr, "Cannot read %s\n", argv[1]); return 1; }

    /* Split input into individual trees (separated by ROOT) */
    char *p = text;
    int tree_count = 0;
    while (*p) {
        /* find next (ROOT */
        char *root = strstr(p, "(ROOT");
        if (!root) break;
        /* find matching close paren */
        int depth = 0;
        char *end = root;
        while (*end) {
            if (*end == '(') depth++;
            else if (*end == ')') { depth--; if (depth == 0) { end++; break; } }
            end++;
        }

        tree_t *t = parse_tree(root);
        if (t && t->n_nodes > 0) {
            tree_count++;
            for (int i = 0; i < t->n_nodes; i++) {
                if (!n_is_cd_id(t, i)) continue;
                char defn[512];
                int dlen = match_definition(t, i, defn, sizeof(defn));
                char *tid = n_text(t, i);
                char *math = idmap_lookup(idmap, tid);
                if (!math) math = tid;
                printf("%s\t%s\t%s\n", tid, math,
                       dlen >= 3 ? defn : "");
            }
        }
        free_tree(t);
        p = end;
    }

    free(text);
    free_idmap(idmap);
    return 0;
}

File: tools/extract_defs.py
"""
Extract mathematical variable definitions from parse trees.
Uses constituency parse trees from meta/profile parser.
Input: .parsed.txt file with S-expression trees
Output: (token_id, math_text, definition) triples

Algorithm: for each (CD <id>) leaf, find the smallest article-starting
NP containing it, and collect all words from that NP minus the ID.
"""
import re, sys

def parse_parsed_file(path):
    """Read .parsed.txt, return list of S-expression tree strings."""
    with open(path) as f:
        text = f.read()
    trees, depth, start = [], 0, -1
    for i, c in enumerate(text):
        if c == '(':
            if depth == 0: start = i
            depth += 1
        elif c == ')':
            depth -= 1
            if depth == 0 and start >= 0:
                trees.append(text[start:i+1])
                start = -1
    return trees


def parse_tree(s, i=0):
    """Parse one S-expression tree into ('N', label, [children]) or ('T', tag, word)."""
    while i < len(s) and s[i] in ' \n\t\r': i += 1
    if i >= len(s) or s[i] != '(': return None, i
    i += 1
    while i < len(s) and s[i] in ' \n\t\r': i += 1
    j = i
    while j < len(s) and s[j] not in ' \n\t\r()': j += 1
    label = s[i:j]; i = j
    children = []
    while i < len(s):
        while i < len(s) and s[i] in ' \n\t\r': i += 1
        if i >= len(s): break
        if s[i] == ')': i += 1; break
        if s[i] == '(':
            child, i = parse_tree(s, i)
            if child: children.append(child)
        else:
            j = i
            while j < len(s) and s[j] not in ' \n\t\r()': j += 1
            children.append(('T', label, s[i:j])); i = j
    return ('N', label, children), i


def contains_id(node, tid):
    if node[0] == 'T': return node[2] == tid
    return any(contains_id(c, tid) for c in node[2])


PUNCT = {'.',',',':',';','--','-','(',')',"''",'``','$','#',
         "n't","'s","'ll","'ve","'re",'...','%','&','='}


def get_words(node, exclude=None):
    """Get all English words from a subtree, skipping IDs and punctuation."""
    if node[0] == 'T':
        w = node[2]
        if w == exclude: return []
        if w.isdigit(): return []
        if w in PUNCT: return []
        return [w]
    words = []
    for c in node[2]:
        words.extend(get_words(c, exclude))
    return words


def find_ids(node):
    """Find all numeric token IDs in a tree."""
    ids = []
    def walk(n):
        if n[0] == 'T' and n[2].isdigit(): ids.append(n[2])
        elif n[0] == 'N':
            for c in n[2]: walk(c)
    walk(node)
    return ids


def np_phrase(np_node, tid):
    """Extract the defining phrase from an NP node containing tid.
    Collects all words from the NP, skipping the ID itself."""
    words = []
    for c in np_node[2]:
        if contains_id(c, tid):
            if c[0] == 'N':
                words.extend(get_words(c, tid))
        else:
            words.extend(get_words(c, tid))
    return ' '.join(words).strip()


def extract_definition(root, tid):
    """Find the best definition NP for a token ID.
    
    Returns the smallest article-starting NP phrase containing the ID.
    If no article-starting NP exists, returns the smallest NP with 3+ words.
    """
    candidates = []
    
    def search(n, depth=0):
        if n[0] != 'N': return
        if n[1] == 'NP' and any(contains_id(c, tid) for c in n[2]):
            phrase = np_phrase(n, tid)
            if len(phrase) >= 3:
                starts_art = phrase.lower().startswith(('the ', 'a ', 'an '))
                wc = len(phrase.split())
                candidates.append((not starts_art, wc, -depth, phrase))
        for c in n[2]:
            search(c, depth + 1)
    
    search(root)
    if not candidates: return None
    
    # Prefer: starts with article, then smallest word count, then deepest
    candidates.sort()
    return candidates[0][3]


def extract_all(trees, id_info=None):
    """Extract definitions from all parse trees.
    id_info: optional dict of token_id -> (type, text) for display.
    Returns list of (token_id, definition) tuples.
    """
    results = []
    for tree_text in trees:
        parsed, _ = parse_tree(tree_text)
        if not parsed: continue
        for tid in find_ids(parsed):
            defn = extract_definition(parsed, tid)
            if defn:
                results.append((tid, defn))
    return results


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python extract_defs.py <file.parsed.txt> [sidecar.tok]")
        sys.exit(1)
    
    trees = parse_parsed_file(sys.argv[1])
    
    # Optional: load token info lookup
    id_info = {}
    if len(sys.argv) > 2:
        with open(sys.argv[2]) as f:
            for line in f:
                m = re.search(r'token_id:(\d+).*?type:([^,]+).*?<tok:(.*?)>', line)
                if m: id_info[m.group(1)] = (m.group(2), m.group(3).strip())
    
    for tid, defn in extract_all(trees, id_info):
        ttype, text = id_info.get(tid, ('?', '?'))
        print(f'{tid} {text[:50]:50s} {defn}')

